odoo/tools/sql.py

```diff
@@ -79,73 +79,78 @@ class SQL:
     its value is a field which the SQL code depends on.  The metadata of a
     wrapper and its parts can be accessed by the iterator ``sql.to_flush``.
     """
-    __slots__ = ('__code', '__args', '__to_flush')
+    __slots__ = ('__code', '__params', '__to_flush')
+
+    __code: str
+    __params: tuple
+    __to_flush: tuple
 
     # pylint: disable=keyword-arg-before-vararg
-    def __new__(cls, code: (str | SQL) = "", /, *args, to_flush: (Field | None) = None, **kwargs):
+    def __init__(self, code: (str | SQL) = "", /, *args, to_flush: (Field | None) = None, **kwargs):
         if isinstance(code, SQL):
-            return code
+            if args or kwargs:
+                raise TypeError("SQL() unexpected arguments when code has type SQL")
+            self.__code = code.__code
+            self.__params = code.__params
+            self.__to_flush = code.__to_flush
+            return
 
         # validate the format of code and parameters
         if args and kwargs:
             raise TypeError("SQL() takes either positional arguments, or named arguments")
+
         if args:
             code % tuple("" for arg in args)
         elif kwargs:
             code, args = named_to_positional_printf(code, kwargs)
-
-        self = object.__new__(cls)
-        self.__code = code
-        self.__args = args
-        self.__to_flush = to_flush
-        return self
-
-    @property
-    def to_flush(self) -> Iterator[Field]:
-        """ Return an iterator on the fields to flush in the metadata of
-        ``self`` and all of its parts.
-        """
-        for node in self.__postfix():
-            if isinstance(node, SQL) and node.__to_flush is not None:
-                yield node.__to_flush
+        else:
```

Comment:
Even for the else case, you should check the code string as the caller may have forgotten to add parameters.
You probably should have:

```
if kwargs:
  code, args = ....
else:
  # just check the format
  code % ([""] * len(args))
```